compiler/test/dotty/tools/backend/jvm/PublicInBinaryTests.scala

```diff
@@ -0,0 +1,342 @@
+        |  @publicInBinary protected given protectedGiven2(using Int): Int = 1
+        |  inline def inlined =
+        |    packagePrivateGiven1 + protectedGiven1 + packagePrivateGiven2(using 1) + protectedGiven2(using 1)
+        |  def testInlined = inlined
+      """.stripMargin
+    checkBCode(code) { dir =>
+      val cClass = loadClassNode(dir.lookupName("C.class", directory = false).input, skipDebugInfo = false)
+      checkPublicMethod(cClass, "packagePrivateGiven1", "()I")
+      checkPublicMethod(cClass, "protectedGiven1", "()I")
+
+      checkPublicMethod(cClass, "packagePrivateGiven2", "(I)I")
+      checkPublicMethod(cClass, "protectedGiven2", "(I)I")
+
+      // Check that the @publicInBinary annotated method is called
+      val testInlined = getMethod(cClass, "testInlined")
+      val testInlinedInstructions = instructionsFromMethod(testInlined).filter(_.isInstanceOf[Invoke])
+      assertSameCode(testInlinedInstructions, List(
+        Invoke(INVOKEVIRTUAL, "C", "packagePrivateGiven1", "()I", false),
+        Invoke(INVOKEVIRTUAL, "C", "protectedGiven1", "()I", false),
+        Invoke(INVOKEVIRTUAL, "C", "packagePrivateGiven2", "(I)I", false),
+        Invoke(INVOKEVIRTUAL, "C", "protectedGiven2", "(I)I", false),
+      ))
+    }
+  }
+
+  @Test
+  def publicInBinaryClassParam(): Unit = {
+    val code =
+      """import scala.annotation.publicInBinary
+        |class C(
+        |  @publicInBinary private[C] val packagePrivateVal: Int = 1,
+        |  @publicInBinary protected val protectedVal: Int = 1,
+        |) {
+        |  inline def inlined =
+        |    packagePrivateVal + protectedVal
+        |  def testInlined = inlined
+        |}
+      """.stripMargin
+    checkBCode(code) { dir =>
+      val cClass = loadClassNode(dir.lookupName("C.class", directory = false).input, skipDebugInfo = false)
+      checkPublicMethod(cClass, "packagePrivateVal", "()I")
+      checkPublicMethod(cClass, "protectedVal", "()I")
+
+      // Check that the @publicInBinary annotated method is called
+      val testInlined = getMethod(cClass, "testInlined")
+      val testInlinedInstructions = instructionsFromMethod(testInlined).filter(_.isInstanceOf[Invoke])
+      assertSameCode(testInlinedInstructions, List(
+        Invoke(INVOKEVIRTUAL, "C", "packagePrivateVal", "()I", false),
+        Invoke(INVOKEVIRTUAL, "C", "protectedVal", "()I", false),
+      ))
+    }
+  }
+
+  @Test
+  def publicInBinaryObject(): Unit = {
+    val code =
+      """package foo
+        |import scala.annotation.publicInBinary
+        |@publicInBinary private object PrivateObject
+        |@publicInBinary private[foo] object PackagePrivateObject
+        |@publicInBinary protected object ProtectedObject
+      """.stripMargin
+    checkBCode(code) { dir =>
+      val privateObject = loadClassNode(dir.subdirectoryNamed("foo").lookupName("PrivateObject$.class", directory = false).input, skipDebugInfo = false)
+      checkPublicField(privateObject, "MODULE$")
```

Comment:
Could we check that the class itself is public too?

Reply:
Yes